dex-it/src/test/scala/com/wavesplatform/it/sync/MarketOrderTestSuite.scala

```diff
@@ -1,118 +1,429 @@
-      node.reservedBalance(bob) shouldBe empty
+      val orderBook = node.orderBook(wavesUsdPair)
+      orderBook.asks should have size 1
+      orderBook.bids should be(empty)
 
-      node.orderBook(wavesUsdPair).asks shouldBe empty
-      node.orderBook(wavesUsdPair).bids shouldBe List { LevelResponse(1.waves, 1.21.usd) }
+      node.accountBalances(account.toAddress.toString)._1 should be(marketOrderAmount - fixedFee)
+      node.assetBalance(account.toAddress.toString, UsdId.toString).balance should be(
+        accountUsdBalance - 5 * 0.2.usd - 15 * 0.3.usd - 30 * 0.4.usd - 100 * 0.5.usd)
     }
   }
 
-  "Market order should be executed correctly when available for spending < required by spendable asset" in {
-    pending
+  "Validation market orders" - {
+    val price = 0.4.usd
+
+    "should be accepted if there is no way to fill it completely (sum of all orders in order book < amount of market order" in {
+      placeOrders(bob, wavesUsdPair, SELL)(100.waves -> price)
+
+      val total = node.orderBook(wavesUsdPair).bids.map(_.amount).sum
+
+      node.waitOrderProcessed(wavesUsdPair,
+                              node.placeMarketOrder(node.prepareOrder(alice, wavesUsdPair, BUY, total + 1000.waves, price, fixedFee)).message.id)
+    }
+
+    "should be accepted if user doesn't have enough Waves to pay fee, but he take waves from order result " in {
+      val amount   = 100.waves
+      val price    = 0.1.usd
+      val transfer = 100.usd
+
+      val account = createAccountWithBalance {
+        transfer -> Some(UsdId.toString)
+      }
+
+      placeOrders(alice, wavesUsdPair, SELL)(
+        amount -> price
+      )
+
+      node.waitOrderProcessed(wavesUsdPair, node.placeMarketOrder(node.prepareOrder(account, wavesUsdPair, BUY, amount, price, fixedFee)).message.id)
+
+      node.accountBalances(account.toAddress.toString)._1 should be(amount - fixedFee)
+    }
+
+    "should be rejected if the price is too low for completely filling by current opened orders (BUY)" in {
+      val amount      = 100.waves
+      val marketPrice = price - 0.1.usd
+
+      placeOrders(bob, wavesUsdPair, SELL)(amount -> price)
+
+      assertBadRequestAndMessage(node.placeMarketOrder(node.prepareOrder(alice, wavesUsdPair, BUY, amount, marketPrice, fixedFee)),
+                                 tooLowPrice("buy", "0.3"))
+    }
+
+    "should be rejected if the price is too high for completely filling by current opened orders (SELL)" in {
+      val amount      = 100.waves
+      val marketPrice = price + 0.1.usd
+
+      placeOrders(bob, wavesUsdPair, BUY)(amount -> price)
+
+      assertBadRequestAndMessage(node.placeMarketOrder(node.prepareOrder(alice, wavesUsdPair, SELL, amount, marketPrice, fixedFee)),
+                                 tooHighPrice("sell", "0.5"))
+    }
+
+    "should be rejected if amount of the buy market order more then user could buy" in {
+      val amount   = 101.waves
```

Comment:
change test name, add "but has not enough balance to pay fee "

Reply:
ok